astropy/units/format/generic.py

```diff
@@ -644,10 +644,12 @@ def _decompose_to_known_units(cls, unit: CompositeUnit | NamedUnit) -> UnitBase:
         are "known" to a given format.
         """
         if isinstance(unit, core.CompositeUnit):
-            new_unit = core.Unit(unit.scale)
-            for base, power in zip(unit.bases, unit.powers):
-                new_unit = new_unit * cls._decompose_to_known_units(base) ** power
-            return new_unit
+            return core.CompositeUnit(
+                unit.scale,
+                [cls._decompose_to_known_units(base) for base in unit.bases],
```

Comment:
This doesn't necessarily seem the same as before. Consider decomposing from SI to natural units, the bases mix. That should be covered in the above, but here can't we get units of the same dimension type (different powers) spread across the list comprehension?
e.g. [m/s, m^3/kg] 
Happy to be wrong about this!

Reply:
> Consider decomposing from SI to natural units...

Why would this function ever do that?

Reply:
That was an illustrative example. The point was that isn't list comprehension different that what was happening before? Or will it always end up being the same?

Reply:
Does this answer your question?
```python
>>> from astropy import units as u
>>> u.CompositeUnit(1, [u.m / u.s], [2])
Unit("m2 / s2")
```

Reply:
It is not the same procedure, and I had to think about it too, but it does do the same thing: here, the new approach just does in one fell swoop what the old loop did in pieces, by basically creates a new `Composite` unit for each base that is added -- `__mul__(self, m)` just does `CompositeUnit(1, [self, m], [1, 1], _error_check=False)`.